### PR TITLE
Implemented MediaElement ObjectDisposedException fix

### LINKED ISSUE #1571: [Bug] MediaElement crashes the program if navigated away from too quickly
### Description
When using MediaElement in a Xamarin Forms application, if the page is navigated away from before the player finishes setting up, it will throw the exception: System.ObjectDisposedException: 'Cannot access a disposed object.
Object name: 'Xamarin.CommunityToolkit.UI.Views.FormsVideoView'.' This is when using a remote source for the video.
I've also seen this issue happen when the page with the video is navigated away from before the video with a remote source starts playing, but I've been unable to reliably reproduce that.
Version tested was Android. Unknown if error exists in IOS as well.

### Steps to Reproduce

1. Download the repro
2. Click the open video button
3. Click the back button before the black box of the video player appears

### Expected Behavior
Page is navigated away from successfully.
### Actual Behavior
App crashes.
### Basic Information

- Version with issue: 1.2.0
- IDE: Visual Studio Community 2019
- Platform Target Frameworks: 
  - iOS:  Xamarin default
  - Android: 11.0
- Nuget Packages:
NETStandard.Library
Xamarin.CommunityToolkit
Xamarin.Essentials
Xamarin.Forms
- Affected Devices:
Motorola One 5G Ace - Android 10

### Reproduction Link

https://github.com/NickCRowork/VideoViewError

### LINKED ISSUE #1769: [Bug] MediaElement URI source throws ObjectDisposedException if return to previous page too early
### Description

On Android, when using MediaElement URI source, there is a possibility that extracting metadata is taking a little longer, when it did taking a little longer and the MediaElementRenderer is already disposed by navigating to the previous page there will be ObjectDisposedException thrown

### Stack Trace

```
  at Java.Interop.JniPeerMembers.AssertSelf (Java.Interop.IJavaPeerable self) [0x00029] in <00c315a988634383b446eff646084784>:0 
  at Java.Interop.JniPeerMembers+JniInstanceMethods.InvokeVirtualVoidMethod (System.String encodedMember, Java.Interop.IJavaPeerable self, Java.Interop.JniArgumentValue* parameters) [0x00000] in <00c315a988634383b446eff646084784>:0 
  at Android.Widget.VideoView.SetVideoURI (Android.Net.Uri uri, System.Collections.Generic.IDictionary`2[TKey,TValue] headers) [0x00041] in /Users/builder/azdo/_work/1/s/xamarin-android/src/Mono.Android/obj/Release/monoandroid10/android-30/mcw/Android.Widget.VideoView.cs:816 
  at Xamarin.CommunityToolkit.UI.Views.FormsVideoView.<>n__1 (Android.Net.Uri uri, System.Collections.Generic.IDictionary`2[TKey,TValue] headers) [0x00000] in <38847864b0e74b799fdf2eaad0701aed>:0 
  at Xamarin.CommunityToolkit.UI.Views.FormsVideoView.SetVideoURI (Android.Net.Uri uri, System.Collections.Generic.IDictionary`2[TKey,TValue] headers) [0x0007d] in <38847864b0e74b799fdf2eaad0701aed>:0 
  at System.Runtime.CompilerServices.AsyncMethodBuilderCore+<>c.<ThrowAsync>b__7_0 (System.Object state) [0x00000] in /Users/builder/jenkins/workspace/archive-mono/2020-02/android/release/mcs/class/referencesource/mscorlib/system/runtime/compilerservices/AsyncMethodBuilder.cs:1021 
  at Android.App.SyncContext+<>c__DisplayClass2_0.<Post>b__0 () [0x00000] in /Users/builder/azdo/_work/1/s/xamarin-android/src/Mono.Android/Android.App/SyncContext.cs:36 
  at Java.Lang.Thread+RunnableImplementor.Run () [0x00008] in /Users/builder/azdo/_work/1/s/xamarin-android/src/Mono.Android/Java.Lang/Thread.cs:36 
  at Java.Lang.IRunnableInvoker.n_Run (System.IntPtr jnienv, System.IntPtr native__this) [0x00008] in /Users/builder/azdo/_work/1/s/xamarin-android/src/Mono.Android/obj/Release/monoandroid10/android-30/mcw/Java.Lang.IRunnable.cs:84 
  at (wrapper dynamic-method) Android.Runtime.DynamicMethodNameCounter.39(intptr,intptr)
```

### Link to Reproduction Sample

https://github.com/yunusefendi52/AppFixCommunityToolkitMedia

### Steps to Reproduce

It's not a "one-liner" reproduce, but please see the reproduction sample above

### Expected Behavior

Not crash ObjectDisposedException

### Actual Behavior

crash ObjectDisposedException

### Basic Information

- Version with issue: 1.3.1
- Last known good version: none
- IDE: VS for Mac
- Platform Target Frameworks: 
  - iOS:  
  - Android: Android 11 (API Level 30)
  - UWP:   
- Android Support Library Version: 
- Nuget Packages: https://www.nuget.org/packages/Xamarin.CommunityToolkit/1.3.1
- Affected Devices:

### Workaround
Uses reflection to invoke base base method ([see repro above in AppFormsVideoView.cs](https://github.com/yunusefendi52/AppFixCommunityToolkitMedia/blob/f1ac2df1edea6a2f6fd3195cb7f15458a002f178/AppCommunityToolkitMedia.Android/AppFormsVideoView.cs#L43))

### Fixing
I think the fix is to check inside SetVideoURI (FormsVideoView.android.cs) if it's disposed or not. I don't mind to create PR for this if you think it is the fix

### PR DESCRIPTION
**Pull request by anvuks**
### Description of Bug ###

FormsVideoView instance was never set to null after disposing, which would sometimes lead to accessing the disposed object and throwing ObjectDisposedException.
I have also rearranged method calls in [SetVideoURI](https://github.com/xamarin/XamarinCommunityToolkit/blob/22b1d4ccf45555c2ed89ecb94e9ff93a5cf61ab3/src/CommunityToolkit/Xamarin.CommunityToolkit/Views/MediaElement/Android/FormsVideoView.android.cs#L53) to avoid calling the base method after the await, during which the instance could get disposed because the overridden method is async void. The only downside of this approach I could think of is that the metadata (video size and duration) may update after the video has already started. Maybe we could revert this and override Dispose and set a private field `isDisposed = true` and then after the await check if `isDisposed == false` to call the base method?

### Issues Fixed ###

- Fixes #1769
- Fixes #1571

### PR Checklist ###

- [x] Has a linked Issue, and the Issue has been `approved`
- [ ] Has tests (if omitted, state reason in description)
- [ ] Has samples (if omitted, state reason in description)
- [x] Rebased on top of main at time of PR
- [x] Changes adhere to coding standard
- [ ] Updated [documentation](https://github.com/MicrosoftDocs/xamarin-communitytoolkit)

### CODE REVIEW

File: src/CommunityToolkit/Xamarin.CommunityToolkit/Views/MediaElement/Android/FormsVideoView.android.cs
Comment (pictos) [inline src/CommunityToolkit/Xamarin.CommunityToolkit/Views/MediaElement/Android/FormsVideoView.android.cs:55]: if we cache the `url` in a local variable, does it work?

```csharp
var localUri = uri;
// rest of the code use `localUri`
```

Comment (anvuks) [reply src/CommunityToolkit/Xamarin.CommunityToolkit/Views/MediaElement/Android/FormsVideoView.android.cs:55]: It should work without any issues, but can you elaborate why we should do that? The call to the base method throws an exception (if the object is disposed), so it wouldn't matter if we cache the argument.

Comment (pictos) [reply src/CommunityToolkit/Xamarin.CommunityToolkit/Views/MediaElement/Android/FormsVideoView.android.cs:55]: That will avoid the usage of a disposed member. I looked into the SetMetadata method and can't see any dispose logical there... So it's something else that's disposing the URL, probably because we're using an async void and it's not waited.

Comment (anvuks) [reply src/CommunityToolkit/Xamarin.CommunityToolkit/Views/MediaElement/Android/FormsVideoView.android.cs:55]: Take a look [here](https://github.com/xamarin/XamarinCommunityToolkit/blob/22b1d4ccf45555c2ed89ecb94e9ff93a5cf61ab3/src/CommunityToolkit/Xamarin.CommunityToolkit/Views/MediaElement/Android/MediaElementRenderer.android.cs#L339). The uri that is passed down to `SetVideoURI()` is never disposed. To elaborate, MediaElement is working correctly unless you close the parent page while it's loading metadata, which will `Dispose()` the renderer and in turn dispose `FormsVideoView`. In the old implementation, metadata awaiting was before the base method call. If the user closes the page while we are awaiting metadata updates, then the next call to base.SetVideoURI will throw and since we are in a async void method we can't catch any exceptions outside. Of course, we can do this after awaiting:

```suggestion
try
{
  base.SetVideoURI(uri, headers);
}
catch(ObjectDisposedException) { }
```
but that is bad code in my opinion.

I caught this bug by accident, by opening and closing the page really fast. After rearranging them, there were no more exceptions. The other bug (not set to null after disposing) was more frequent since the `MetadataRetrieved` event was raised before the base call and the renderer tried to access the disposed instance first.

**Comment by pictos** [comment]
@anvuks before we merge this you need to assign the CLA, can you do that please?

File: src/CommunityToolkit/Xamarin.CommunityToolkit/Views/MediaElement/Android/FormsVideoView.android.cs
Comment (pictos) [reply src/CommunityToolkit/Xamarin.CommunityToolkit/Views/MediaElement/Android/FormsVideoView.android.cs:55]: @anvuks I believe that your PR fixes the issue. I just want to make sure that your PR doesn't cause any side effects. I'll test it tomorrow and merge it if everything works during my test. Also, I invite @peterfoot to take a look on this if he can

**Comment by anvuks** [comment]
@pictos I already did when I submitted this PR, but for some reason the bot says it's not signed. Does it usually take this long or do I have to do something else? I've also clicked on recheck link but nothing happens.

**Comment by pictos** [comment]
@anvuks sometimes it gets crazy hehe
I'll ask someone on the top check if everything is good.

**Comment by anvuks** [comment]
I have closed and reopened this PR to try to fix pending cla issue [by triggering the PR webhook](https://github.com/cla-assistant/cla-assistant/issues/441#issuecomment-495528696) without any luck and just noticed that [microsoft-cla](https://github.com/apps/microsoft-cla) bot's [comment](https://github.com/xamarin/XamarinCommunityToolkit/pull/1870#issuecomment-1146402985) has been overwritten by the [net-foundation-cla](https://github.com/apps/net-foundation-cla) bot with the same `CLA Signed` message. I don't know how it missed, but there is nothing I can do about it besides opening another PR. 

If it's okay with you, can you merge this PR?

**Comment by jfversluis** [comment]
As far as the CLA goes: one is green so that should be all good! :)

**Review by brettnguyen — APPROVED** [review]

**Comment by brettnguyen** [comment]
So when will this be approved to be merged?

**Comment by brettnguyen** [comment]
> I have closed and reopened this PR to try to fix pending cla issue [by triggering the PR webhook](https://github.com/cla-assistant/cla-assistant/issues/441#issuecomment-495528696) without any luck and just noticed that [microsoft-cla](https://github.com/apps/microsoft-cla) bot's [comment](https://github.com/xamarin/XamarinCommunityToolkit/pull/1870#issuecomment-1146402985) has been overwritten by the [net-foundation-cla](https://github.com/apps/net-foundation-cla) bot with the same `CLA Signed` message. I don't know how it missed, but there is nothing I can do about it besides opening another PR.
> 
> If it's okay with you, can you merge this PR?



> I have closed and reopened this PR to try to fix pending cla issue [by triggering the PR webhook](https://github.com/cla-assistant/cla-assistant/issues/441#issuecomment-495528696) without any luck and just noticed that [microsoft-cla](https://github.com/apps/microsoft-cla) bot's [comment](https://github.com/xamarin/XamarinCommunityToolkit/pull/1870#issuecomment-1146402985) has been overwritten by the [net-foundation-cla](https://github.com/apps/net-foundation-cla) bot with the same `CLA Signed` message. I don't know how it missed, but there is nothing I can do about it besides opening another PR.
> 
> If it's okay with you, can you merge this PR?

when will this be merged?

**Comment by anvuks** [comment]
@brettnguyen It should be merged as soon as a maintainer approves my changes. If this fix is critical to your app, you are going to have to use a nightly release until the next stable release.

@pictos CLA issues have been resolved, so if you have the time can you merge this please? 😁

**Comment by pictos** [comment]
@anvuks I'll take a look this week asap

**Review by pictos — APPROVED** [review]
@anvuks thanks for this ❣️ And sorry for the long time to review and approve this. 
I hope this doesn't discourage you from making more contributions.

**Comment by anvuks** [comment]
@pictos Thanks! Don't worry about it, I know the CommunityToolkit maintainers are doing it all in their free time and life sometimes gets hectic. I'll be contributing more for sure 😀

File: src/CommunityToolkit/Xamarin.CommunityToolkit/Views/MediaElement/Android/FormsVideoView.android.cs
Comment (H7mosh) [reply src/CommunityToolkit/Xamarin.CommunityToolkit/Views/MediaElement/Android/FormsVideoView.android.cs:55]: @pictos  @anvuks  can you guys tell me where can i find the file that need to be edited by adding try and catch ???

**Pull request merged**
